txpipe/magnification.py

```diff
@@ -0,0 +1,204 @@
+from .base_stage import PipelineStage
+from .data_types import (
+    YamlFile,
+    HDFFile,
+    FitsFile,
+    PNGFile,
+)
+from .utils import LensNumberDensityStats, Splitter, rename_iterated
+from .binning import build_tomographic_classifier, apply_classifier
+import numpy as np
+import warnings
+
+
+class TXSSIMagnification(PipelineStage):
+    """
+    class for computing the magnification coefficients using SSI outputs
+    Following the methodology of https://arxiv.org/abs/2012.12825
+    and https://arxiv.org/abs/2209.09782
+    """
+
+    name = "TXSSIMagnification"
+
+    inputs = [
+        ("binned_lens_catalog_nomag", HDFFile),
+        ("binned_lens_catalog_mag", HDFFile),
+    ]
+
+    outputs = [
+        ("magnification_coefficients", HDFFile),
+        ("magnification_plot", PNGFile),
+    ]
+
+    config_options = {
+        "chunk_rows": 10000,
+        "applied_magnification": 1.02, #magnification applied to the "magnified" SSI catalog
+        #TO DO: add a way for "applied_magnification" to be determined from the SSI inputs directly
+        "n_patches":20,
+        "bootstrap_error":True,
+    }
+
+    def run(self):
+        """
+        Run the analysis for this stage.
+        """
+        from scipy.stats import bootstrap
+
+        # load the catalogs
+        nomag_cat = self.open_input("binned_lens_catalog_nomag")
+        mag_cat   = self.open_input("binned_lens_catalog_mag")
+
+        # get/estimate the magnification applied to the magnified catalog
+        # TO DO: could put this as optional config item in TXSSIIngest
+        # so that it is done automatically
+        mu = self.config["applied_magnification"]
+        deltak = (1. - 1./mu)/2.
+        
+        if self.config["bootstrap_error"]:
+            #split no-mag sample up into patches to be used in the bootstrap errors
+            cluster = self.calc_cluster_patches(nomag_cat)
+
+        # compute fractional change in number count for both samples
+        # in each lens bin
+        nbins = nomag_cat['lens/'].attrs['nbin_lens']
+        outfile = self.setup_output(nbins)
+        for ibin in range(nbins):
```

Comment:
Do we also want to run this on the non-tomographic ("2D") sample?

Reply:
yes good idea. i added an extra "bin" index that will contain the full sample `bin_all`